edu-services/cm-service/cm-impl/hibernate-impl/impl/src/java/org/sakaiproject/coursemanagement/impl/CourseSitePublishServiceImpl.java

```diff
@@ -0,0 +1,244 @@
+    *             if this method is invoked at any time during this
+    *             time period, unpublished course sites associated
+    *             with this term will be published.
+    *
+    * </p>
+    * <p>
+    * There is a special use case to consider.
+    * Say that this method is invoked, and all the course sites for the current term are published.
+    * Now, suppose one particular instructor forgot to add some content, or made some mistakes setting up his course site, and he wants to unpublish his couse site while
+    * he makes some changes to it.  He goes into sakai, and uses the site setup tool to mark the course site as unpublished.  But, suppose the next day this service is
+    * invoked again (via a quartz job).  His course would get published again, even though he doesn't want it to be, and has manually gone into sakai and set it to
+    * unpublished.  In order to prevent this case from happening, whenever this service marks a course site as published, it also sets a site property called <i>course_site_publish_service.publish.set</i>.
+    * Then, if an instructor manually resets his site to be unpublished, this service will not automatically publish it.  Rather, the instructor who manually unpublished the
+    * course site will have to go into sakai, use the site setup tool, and manually publish it.
+    * </p>
+    * </br></br>
+    * @param numDaysBeforeTermStarts   number of days before a term starts that course sites should be published.
+    * </br></br>
+    * @return the number of course sites that were published.
+    */
+   public int publishCourseSites(int numDaysBeforeTermStarts) {
+
+      logger.info("publishCourseSites(" + numDaysBeforeTermStarts + " days before the term starts)");
+
+      Date today             = new Date();
+      Date publishDate       = null;
+      int  numSitesPublished = 0;
+
+      try {
+         // get the list of the academic terms
+         List<AcademicSession> academicSessions = courseManagementService.getAcademicSessions();
+
+         for(AcademicSession academicSession : academicSessions) {
+            publishDate = new Date(academicSession.getStartDate().getTime() - numDaysBeforeTermStarts * ONE_DAY_IN_MS);
+            // see if the academic is scheduled to start within the specified number of days
+            if (publishDate.getTime() < today.getTime() && today.getTime() < academicSession.getEndDate().getTime()) {
+
+               // get a list of all published and unpublished course sites in ascending creation date order which are associated with the specified academic session
+               Hashtable<String, String> propertyCriteria = new Hashtable<String, String>();
+               propertyCriteria.put("term_eid", academicSession.getEid());
+               List<Site> sites = (List<Site>)siteService.getSites(SelectionType.NON_USER, "course", null, propertyCriteria, SortType.CREATED_ON_ASC, null);
```

Comment:
I agree that having an option in the SiteService API that allows to search for only published sites or unpublished sites will certainly decrease the number of results and make lighter the job. 

The actual selectionType in SiteService allows ignoreSpecial, ignoreUser, ignoreUnpublished (that can be useful for the unpublish job) and ignore softlyDeleted)... we should add an ignorePublished there, and then create a new selectioType (and modify all the others with ignorePublished=false)...

I think it is not really complex, (we need to follow the same work done with the soft_deletion filter https://crucible.sakaiproject.org/changelog/Sakai.Git?cs=7307b17239f57a80217876e1fcf110806b750ec0)
Here is the JIRA. I've created it in KNL, but maybe I should had done that in SAK

https://jira.sakaiproject.org/browse/KNL-1393

Of course, if you consider this correct, I'd like to make this actual pull request first, then solve the KNL-1393, and then return to update this part of the code with the new getSites search.